### ISSUE TITLE
Pylint does not appear to understand the descriptor protocol

### ISSUE THREAD

**Issue by tomkcook** [PyCQA/pylint#7820]
### Bug description

Writing a custom `@property`-like descriptor causes pylint to issue various false postives.  Example code:

```
"Demonstrate class property problems"

from typing import Callable

class ClassROProperty(property):
    "Decorator class to create a read-only class property"
    def __init__(self, getter:Callable):
        self._getter = getter

    def __get__(self, _, cls):
        return self._getter(cls)

class MyClass:
    "A class which has only class properties"

    @ClassROProperty
    def prop_1(cls):
        "Return an iterable value"
        return [1, 2, 3, 4]

for val in MyClass.prop_1:
    print(val)
print(MyClass.prop_1[1])
```

Running `pylint --disable=too-few-public-methods` on this results in these failures:

```
************* Module test
test.py:17:4: E0213: Method 'prop_1' should have "self" as first argument (no-self-argument)
test.py:21:11: E1133: Non-iterable value MyClass.prop_1 is used in an iterating context (not-an-iterable)
test.py:23:6: E1136: Value 'MyClass.prop_1' is unsubscriptable (unsubscriptable-object)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

E0213 can be fairly easily handled by disabling `no-self-argument` module- or class-wide.  But E1133 and E1136 are much more difficult to suppress, because they happen everywhere the properties are used, not where they are defined.  So anyone using your code needs to know to suppress these errors in their code.

I raised this on discord and was asked to file a report here.

### Command used

```shell
venv/bin/pylint --disable=too-few-public-methods test.py
```

### Pylint output

```shell
************* Module test
test.py:17:4: E0213: Method 'prop_1' should have "self" as first argument (no-self-argument)
test.py:21:11: E1133: Non-iterable value MyClass.prop_1 is used in an iterating context (not-an-iterable)
test.py:23:6: E1136: Value 'MyClass.prop_1' is unsubscriptable (unsubscriptable-object)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Expected behavior

These appear to be false positives.

### Pylint version

```shell
pylint 2.15.6
astroid 2.12.13
Python 3.10.6 (main, Nov  2 2022, 18:53:38) [GCC 11.3.0]
```

### OS / Environment

Ubuntu 22.04

### Additional dependencies

# Apologies, this wasn't a very clean venv but I don't think any of this is relevant.
astroid==2.12.13
attrs==22.1.0
bidict==0.22.0
certifi==2022.9.24
charset-normalizer==2.1.1
dill==0.3.6
exceptiongroup==1.0.4
idna==3.4
iniconfig==1.1.1
isort==5.10.1
lazy-object-proxy==1.8.0
mccabe==0.7.0
netifaces==0.10.6
packaging==21.3
platformdirs==2.5.4
pluggy==1.0.0
pylint==2.15.6
pyparsing==3.0.9
pytest==7.2.0
python-engineio==4.3.4
python-socketio==5.7.2
requests==2.28.1
tomli==2.0.1
tomlkit==0.11.6
urllib3==1.26.12
wrapt==1.14.1

**Comment by rafaelclp** [comment]
Bumping.

Pylint still doesn't understand descriptors at all (except for the few ones that seem to be hardcoded in pylint, like classmethod, property etc).

**Comment by jacobtylerwalls** [comment]
A from-the-hip naive patch that solves #9832 but causes some failures worth understanding:
```py
diff --git a/astroid/bases.py b/astroid/bases.py
index 4a684cf1..3dfd71f1 100644
--- a/astroid/bases.py
+++ b/astroid/bases.py
@@ -44,9 +44,7 @@ if PY310_PLUS:
 
 # List of possible property names. We use this list in order
 # to see if a method is a property or not. This should be
-# pretty reliable and fast, the alternative being to check each
-# decorator to see if its a real property-like descriptor, which
-# can be too complicated.
+# pretty reliable and fast.
 # Also, these aren't qualified, because each project can
 # define them, we shouldn't expect to know every possible
 # property-like decorator!
@@ -87,6 +85,8 @@ def _is_property(
         if inferred is None or isinstance(inferred, UninferableBase):
             continue
         if isinstance(inferred, nodes.ClassDef):
+            if inferred.lookup("__get__"):
+                return True
             for base_class in inferred.bases:
                 if not isinstance(base_class, nodes.Name):
                     continue
```
